lib/luacrypto.c

```diff
@@ -0,0 +1,438 @@
+	digest = kmalloc(digestsize, gfp);
+	if (!digest)
+	{
+		crypto_free_shash(tfm);
+		return luaL_error(L, "failed to alloc digest buffer");
+	}
+
+	desc = kmalloc(sizeof(*desc) + crypto_shash_descsize(tfm), gfp);
+	if (!desc)
+	{
+		kfree(digest);
+		crypto_free_shash(tfm);
+		return luaL_error(L, "failed to alloc desc");
+	}
+	desc->tfm = tfm;
+
+	ret = crypto_shash_digest(desc, data, datalen, digest);
+	crypto_free_shash(tfm);
+	kfree(desc);
+
+	if (ret)
+	{
+		kfree(digest);
+		return luaL_error(L, "digest failed (%d)", ret);
+	}
+
+	lua_pushlstring(L, (const char *)digest, digestsize);
+	kfree(digest);
+	return 1;
+}
+
+static int luacrypto_hmac(lua_State *L)
+{
+	size_t keylen, datalen;
+	const char *alg = luaL_checkstring(L, 1);
+	const char *key = luaL_checklstring(L, 2, &keylen);
+	const char *data = luaL_checklstring(L, 3, &datalen);
+	gfp_t gfp = lunatik_gfp(lunatik_toruntime(L));
+
+	char hmac_alg[CRYPTO_MAX_ALG_NAME];
+	snprintf(hmac_alg, sizeof(hmac_alg), "hmac(%s)", alg);
+
+	struct crypto_shash *tfm = crypto_alloc_shash(hmac_alg, 0, 0);
+	if (IS_ERR(tfm))
+		return luaL_error(L, "failed to alloc hmac tfm");
+
+	int ret = crypto_shash_setkey(tfm, key, keylen);
+	if (ret)
+	{
+		crypto_free_shash(tfm);
+		return luaL_error(L, "failed to set hmac key (%d)", ret);
+	}
+
+	int digestsize = crypto_shash_digestsize(tfm);
+	u8 *digest = kmalloc(digestsize, gfp);
+	if (!digest)
+	{
+		crypto_free_shash(tfm);
+		return luaL_error(L, "failed to alloc digest buffer");
+	}
+
+	struct shash_desc *desc = kmalloc(sizeof(*desc) + crypto_shash_descsize(tfm), gfp);
+	if (!desc)
+	{
+		kfree(digest);
+		crypto_free_shash(tfm);
+		return luaL_error(L, "failed to alloc desc");
+	}
+	desc->tfm = tfm;
+
+	ret = crypto_shash_digest(desc, data, datalen, digest);
+	crypto_free_shash(tfm);
+	kfree(desc);
+
+	if (ret)
+	{
+		kfree(digest);
+		return luaL_error(L, "hmac failed (%d)", ret);
+	}
+
+	lua_pushlstring(L, (const char *)digest, digestsize);
+	kfree(digest);
+	return 1;
+}
+
+LUNATIK_PRIVATECHECKER(luacrypto_check, struct crypto_aead *);
+
+/*
+* crypto:setkey(key)
+*
+* Sets the key for the crypto transform. For example, for AES-128-GCM,
+* the key length must be 16 bytes. Other algorithms may require different lengths.
+*/
+static int luacrypto_setkey(lua_State *L)
+{
+	struct crypto_aead *tfm = luacrypto_check(L, 1);
+	size_t keylen;
+	const char *key = luaL_checklstring(L, 2, &keylen);
+	int ret;
+
+	ret = crypto_aead_setkey(tfm, key, keylen);
+	if (ret)
+		luaL_error(L, "failed to set key (%d)", ret);
+
+	lua_pushboolean(L, 1);
+	return 1;
+}
+
+/*
+* crypto:setauthsize(tagsize)
+*
+* Sets the size (in bytes) of the authentication tag.
+*/
+static int luacrypto_setauthsize(lua_State *L)
+{
+	struct crypto_aead *tfm = luacrypto_check(L, 1);
+	int tagsize = luaL_checkinteger(L, 2);
+	int ret;
+
+	ret = crypto_aead_setauthsize(tfm, tagsize);
+	if (ret)
+		luaL_error(L, "failed to set authsize (%d)", ret);
+
+	lua_pushboolean(L, 1);
```

Comment:
why do you need this? you are raising an error if it fails, so if it ever returns, it's success, right?